### PR TITLE
Route based links - v0.9.x

### PR DESCRIPTION
**Pull request by lgebhardt**
This is the same basic code from #1246 reworked for the release-0-9 branch. Changes made to #1246 will be brought to this PR.

### All Submissions:

- [ ] I've checked to ensure there aren't other open [Pull Requests](https://github.com/cerebris/jsonapi-resources/pulls) for the same update/change.
- [ ] I've submitted a [ticket](https://github.com/cerebris/jsonapi-resources/issues) for my issue if one did not already exist.
- [ ] My submission passes all tests. (Please run the full test suite locally to cut down on noise from travis failures.)
- [ ] I've used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message or the description.
- [ ] I've added/updated tests for this change.

### New Feature Submissions:

- [ ] I've submitted an issue that describes this feature, and received the go ahead from the maintainers.
- [ ] My submission includes new tests.
- [ ] My submission maintains compliance with [JSON:API](http://jsonapi.org/).

### Bug fixes and Changes to Core Features:

- [ ] I've included an explanation of what the changes do and why I'd like you to include them.
- [ ] I've provided test(s) that fails without the change.

### Reviewer Checklist:
- [ ] Maintains compliance with JSON:API
- [ ] Adequate test coverage exists to prevent regressions

### CODE REVIEW

**Review by dgeb — APPROVED** [review]

File: lib/jsonapi/acts_as_resource_controller.rb [hunk 1]
```diff
@@ -234,7 +234,7 @@ def render_results(operation_results)
       end
 
       render_options[:location] = content[:data]["links"][:self] if (
-        response_doc.status == :created && content[:data].class != Array
+        response_doc.status == :created && content[:data].class != Array && content[:data]["links"]
```
Comment (dgeb) [inline lib/jsonapi/acts_as_resource_controller.rb:237]: This probably isn't strictly necessary, but it shows that `:location` should not be empty:

```suggestion
        response_doc.status == :created && content[:data].class != Array && content[:data]["links"] && content[:data]["links"][:self]
```

File: lib/jsonapi/relationship.rb [hunk 2]
```diff
@@ -70,6 +93,12 @@ def initialize(name, options = {})
         @foreign_key_on = options.fetch(:foreign_key_on, :self)
       end
 
+      def to_s
+        # :nocov:
+        "#{parent_resource}.#{name}(#{belongs_to? ? 'ToOne' : 'BelongsToOne'})"
```
Comment (dgeb) [inline lib/jsonapi/relationship.rb]:
```suggestion
        "#{parent_resource}.#{name}(#{belongs_to? ? 'BelongsToOne' : 'ToOne'})"
```

Not sure why we want to distinguish `ToOne` from `BelongsToOne` at the resource relationship level. But I think the ternary was backwards.

File: lib/jsonapi/resource_serializer.rb [hunk 3]
```diff
@@ -431,19 +439,20 @@ def to_many_linkage(source, relationship)
     def link_object_to_one(source, relationship, include_linkage)
       include_linkage = include_linkage | @always_include_to_one_linkage_data | relationship.always_include_linkage_data
       link_object_hash = {}
-      link_object_hash[:links] = {}
-      link_object_hash[:links][:self] = self_link(source, relationship)
-      link_object_hash[:links][:related] = related_link(source, relationship)
+
+      links = default_relationship_links(source, relationship)
+
+      link_object_hash['links'] = links unless links.blank?
```
Comment (dgeb) [inline lib/jsonapi/resource_serializer.rb]: if `links_object_hash` represents the relationship, then shouldn't it be named `relationship_hash` or similar?

File: lib/jsonapi/acts_as_resource_controller.rb (hunk 1 above)
Comment (lgebhardt) [reply lib/jsonapi/acts_as_resource_controller.rb:237]: This just makes the relationships easier to read in the debugger. The `to_s` isn't called at other times. Yes, the ternary was backwards.

File: lib/jsonapi/resource_serializer.rb (hunk 3 above)
Comment (lgebhardt) [reply lib/jsonapi/resource_serializer.rb]: Fixed

**Pull request merged**